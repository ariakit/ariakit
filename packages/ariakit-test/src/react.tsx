import type { ReactElement } from "react";
import { StrictMode } from "react";
import * as ReactTestingLibrary from "@testing-library/react";
import { flushMicrotasks, nextFrame, wrapAsync } from "./__utils.js";

export * from "./index.js";

export interface RenderOptions
  extends Omit<ReactTestingLibrary.RenderOptions, "queries"> {
  strictMode?: boolean;
}

export async function render(ui: ReactElement, options?: RenderOptions) {
  const wrapper = (props: { children: ReactElement }) => {
    const Wrapper = options?.wrapper;
    const element = Wrapper ? <Wrapper {...props} /> : props.children;
    if (!options?.strictMode) return element;
    return <StrictMode>{element}</StrictMode>;
  };

  return wrapAsync(async () => {
    const { unmount } = ReactTestingLibrary.render(ui, { ...options, wrapper });
    await flushMicrotasks();
    await nextFrame();
    await flushMicrotasks();
    return unmount;
  });
}
