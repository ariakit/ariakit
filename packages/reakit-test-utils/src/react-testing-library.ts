export {
  act,
  fireEvent,
  render,
  wait,
  waitForDomChange,
  waitForElement,
  waitForElementToBeRemoved
} from "@testing-library/react";
