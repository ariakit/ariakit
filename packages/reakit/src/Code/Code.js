import React from "react";
import PropTypes from "prop-types";
import { theme } from "styled-tools";
import styled from "../styled";
import as from "../as";
import Box from "../Box";

const Component = props => {
  if (props.block) {
    return (
      <Box as="pre" {...props}>
        <code className={props.codeClassName}>{props.children}</code>
      </Box>
    );
  }
  const className = props.codeClassName
    ? `${props.className} ${props.codeClassName}`.trim()
    : `${props.className}`.trim();
  return <Box as="code" {...props} className={className} />;
};

const Code = styled(Box)`
  ${theme("Code")};
`;

Code.propTypes = {
  block: PropTypes.bool,
  codeClassName: PropTypes.string
};

Code.defaultProps = {
  opaque: true,
  palette: "background",
  tone: -2
};

export default as(Component)(Code);
