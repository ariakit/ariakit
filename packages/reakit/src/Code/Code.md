Code works in two modes, for inline code and pre-formatted code blocks. Use the `block` prop to change between modes.

```jsx
<Code>Code by itself will be inline</Code>
```

Code in block mode.

```jsx
<Code block>
  {`//for example:
let foo = 99

if (true) {
  let foo = 2
  //logs "2"
  console.log(foo)
}

//logs "99"
console.log(foo)
`}
</Code>
```
