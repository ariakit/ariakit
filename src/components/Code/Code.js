import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { ifProp, prop } from "styled-tools";
import as from "../../enhancers/as";
import Base from "../Base";

const Component = props => {
  if (props.block) {
    return (
      <Base as="pre" {...props}>
        <code className={props.codeClassName}>{props.children}</code>
      </Base>
    );
  }
  const className = props.codeClassName
    ? `${props.className} ${props.codeClassName}`.trim()
    : `${props.className}`.trim();
  return <Base as="code" {...props} className={className} />;
};

const Code = styled(Base)`
  background-color: rgba(0, 0, 0, 0.05);
  font-family: monospace;
  white-space: pre-wrap;
  word-wrap: break-word;
  padding: ${ifProp("block", "0", "0.25em 0.35em")};

  code {
    display: block;
    padding: 1em;
  }

  ${prop("theme.Code")};
`;

Code.propTypes = {
  block: PropTypes.bool,
  codeClassName: PropTypes.string
};

export default as(Component)(Code);
