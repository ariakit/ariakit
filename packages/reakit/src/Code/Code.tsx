import * as React from "react";
import * as PropTypes from "prop-types";
import { theme } from "styled-tools";
import styled from "../styled";
import Box, { BoxProps } from "../Box";

export interface CodeProps extends BoxProps {
  block?: boolean;
  codeClassName?: string;
}

const CodeComponent = (props: CodeProps) => {
  if (props.block) {
    return (
      <Box use="pre" {...props}>
        <code className={props.codeClassName}>{props.children}</code>
      </Box>
    );
  }
  const className = props.codeClassName
    ? `${props.className} ${props.codeClassName}`.trim()
    : `${props.className}`.trim();
  return <Box use="code" {...props} className={className} />;
};

const Code = styled(Box)<CodeProps>`
  ${theme("Code")};
`;

// @ts-ignore
Code.propTypes = {
  block: PropTypes.bool,
  codeClassName: PropTypes.string
};

Code.defaultProps = {
  use: CodeComponent,
  opaque: true,
  palette: "background",
  tone: -2
};

export default Code;
