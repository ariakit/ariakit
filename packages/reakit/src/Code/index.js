import Code from "./Code";

export default Code;
