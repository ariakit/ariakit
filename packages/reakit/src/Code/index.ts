import Code from "./Code";

export * from "./Code";

export default Code;
