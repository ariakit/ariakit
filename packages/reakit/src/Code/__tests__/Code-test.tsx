import * as React from "react";
import { render } from "react-testing-library";
import Code from "../Code";

test("html attrs", () => {
  const { getByTestId } = render(
    <Code id="test" aria-label="test" data-testid="test" />
  );

  expect(getByTestId("test")).toHaveAttribute("id", "test");
  expect(getByTestId("test")).toHaveAttribute("aria-label", "test");
});

test("styled", () => {
  const { container } = render(<Code />);
  expect(container.firstChild).toMatchSnapshot();
});

test("styled block", () => {
  const { container } = render(<Code block />);
  expect(container.firstChild).toMatchSnapshot();
});

test("styled with custom CSS class", () => {
  const cssClassName = "chuck-norris";
  const { getByTestId } = render(
    <Code codeClassName={cssClassName} data-testid="test" />
  );

  expect(getByTestId("test").className).toContain(cssClassName);
});
