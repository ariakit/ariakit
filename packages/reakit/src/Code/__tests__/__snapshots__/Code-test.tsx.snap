// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled 1`] = `
.c0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

<code
  class="c0"
/>
`;

exports[`styled block 1`] = `
.c0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

<pre
  class="c0"
>
  <code />
</pre>
`;
